### ISSUE TITLE
why does textDocument/completion return null result sometimes?

### ISSUE THREAD

**Issue by prabirshrestha** [sourcegraph/javascript-typescript-langserver#64]
here is the log

```
--> {"method":"initialize","jsonrpc":"2.0","id":1,"params":{"capabilities":{},"rootPath":"file:///D:/tmp/helloworld"}}
--> {"method":"textDocument/didClose","jsonrpc":"2.0","params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"}}}
--> {"method":"textDocument/didSave","jsonrpc":"2.0","params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"text":""}}
<-- {"jsonrpc":"2.0","id":1,"result":{"capabilities":{"textDocumentSync":1,"hoverProvider":true,"definitionProvider":true,"referencesProvider":true,"documentSymbolProvider":true,"workspaceSymbolProvider":true,"xworkspaceReferencesProvider":true,"xdefinitionProvider":true,"xdependenciesProvider":true,"completionProvider":{"resolveProvider":false,"triggerCharacters":["."]}}}}
--> {"method":"textDocument/didOpen","jsonrpc":"2.0","params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts","version":1,"languageId":"typescript","text":""}}}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":2,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":1,"line":0}}}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":3,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":2,"line":0}}}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":4,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":4,"line":0}}}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":5,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":5,"line":0}}}
<-- {"jsonrpc":"2.0","id":2,"result":null}
<-- {"jsonrpc":"2.0","id":3,"result":null}
<-- {"jsonrpc":"2.0","id":4,"result":null}
<-- {"jsonrpc":"2.0","id":5,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":6,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":6,"line":0}}}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":7,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":7,"line":0}}}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":8,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":8,"line":0}}}
<-- {"jsonrpc":"2.0","id":6,"result":null}
<-- {"jsonrpc":"2.0","id":7,"result":null}
<-- {"jsonrpc":"2.0","id":8,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":9,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":9,"line":0}}}
<-- {"jsonrpc":"2.0","id":9,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":10,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":11,"line":0}}}
<-- {"jsonrpc":"2.0","id":10,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":11,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":15,"line":0}}}
<-- {"jsonrpc":"2.0","id":11,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":12,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":16,"line":0}}}
<-- {"jsonrpc":"2.0","id":12,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":13,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":18,"line":0}}}
<-- {"jsonrpc":"2.0","id":13,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":14,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":20,"line":0}}}
<-- {"jsonrpc":"2.0","id":14,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":15,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":21,"line":0}}}
<-- {"jsonrpc":"2.0","id":15,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":16,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":22,"line":0}}}
<-- {"jsonrpc":"2.0","id":16,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":17,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":24,"line":0}}}
<-- {"jsonrpc":"2.0","id":17,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":18,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":26,"line":0}}}
<-- {"jsonrpc":"2.0","id":18,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":19,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":27,"line":0}}}
<-- {"jsonrpc":"2.0","id":19,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":20,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":29,"line":0}}}
<-- {"jsonrpc":"2.0","id":20,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":21,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":30,"line":0}}}
<-- {"jsonrpc":"2.0","id":21,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":22,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":1,"line":2}}}
<-- {"jsonrpc":"2.0","id":22,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":23,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":2,"line":2}}}
<-- {"jsonrpc":"2.0","id":23,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":24,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":4,"line":2}}}
<-- {"jsonrpc":"2.0","id":24,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":25,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":5,"line":2}}}
<-- {"jsonrpc":"2.0","id":25,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":26,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":7,"line":2}}}
<-- {"jsonrpc":"2.0","id":26,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":27,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":9,"line":2}}}
<-- {"jsonrpc":"2.0","id":27,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":28,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":11,"line":2}}}
<-- {"jsonrpc":"2.0","id":28,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":29,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":16,"line":2}}}
<-- {"jsonrpc":"2.0","id":29,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":30,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":18,"line":2}}}
<-- {"jsonrpc":"2.0","id":30,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":31,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":20,"line":2}}}
<-- {"jsonrpc":"2.0","id":31,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":32,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":21,"line":2}}}
<-- {"jsonrpc":"2.0","id":32,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":33,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":23,"line":2}}}
<-- {"jsonrpc":"2.0","id":33,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":34,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":24,"line":2}}}
<-- {"jsonrpc":"2.0","id":34,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":35,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":20,"line":2}}}
<-- {"jsonrpc":"2.0","id":35,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":36,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":21,"line":2}}}
<-- {"jsonrpc":"2.0","id":36,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":37,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":22,"line":2}}}
<-- {"jsonrpc":"2.0","id":37,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":38,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":23,"line":2}}}
<-- {"jsonrpc":"2.0","id":38,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":39,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":25,"line":2}}}
<-- {"jsonrpc":"2.0","id":39,"result":null}
--> {"method":"textDocument/completion","jsonrpc":"2.0","id":40,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":26,"line":2}}}
<-- {"jsonrpc":"2.0","id":40,"result":null}
--> {"method":"textDocument/didClose","jsonrpc":"2.0","params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"}}}
```

**Comment by felixfbecker** [comment]
Not spec-compliant

**Comment by alexsaveliev** [comment]
It seems that there is an error indeed - compare the results of completion when file is empty and when file looks for example like `a`. In the first case there are no results, in the second one there are plenty of them. I'll try to address this issue, if possible

netcat file to demonstrate this issue:

```
Content-Length: 114

{"method":"initialize","jsonrpc":"2.0","id":1,"params":{"capabilities":{},"rootPath":"file:///D:/tmp/helloworld"}}Content-Length: 166

{"method":"textDocument/completion","jsonrpc":"2.0","id":2,"params":{"textDocument":{"uri":"file:///D:/tmp/helloworld/hello.ts"},"position":{"character":8,"line":0}}}
```

**Comment by alexsaveliev** [comment]
claimed fixed by #67
